### PR TITLE
feat(fossil_branch): add fossil_branch module

### PR DESCRIPTION
**Pull request by VegardSkui**
#### Description

Add a module to print the active branch of a [Fossil](https://www.fossil-scm.org/) check-out, if the current directory is a Fossil check-out. The module is analogous to the `git_branch` and `hg_branch` modules.

#### Motivation and Context

This is part of adding initial Fossil support to Starship, as this was currently missing. This lets users of Fossil repositories have some of the same functionality available in their check-outs as they already enjoy in their Git repositories. While Fossil users usually don't switch branches for a single directory as often as e.g. Git users, this module provides a nice indicator that the directory is version controlled and would work nicely together with a `fossil_metrics` module. The `fossil_metrics` module will be implemented in a follow-up pull request.

Contributes to the closing of #4036.

#### How Has This Been Tested?

- [x] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [ ] I have tested using **Windows**

Manual tests were ran both with and without Fossil installed and in multiple check-outs. I have also tested the module on FreeBSD.

#### Screenshots (if appropriate):

NA

#### Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

**Review by davidkna — CHANGES_REQUESTED** [review]
I haven't used fossil yet, but it looks like it would be reasonable to scan for `.fslckout` before running the external command. Please also add more tests, including a branch-available test.

**Comment by VegardSkui** [comment]
Agreed. I've added more test in the style of `hg_branch` and `pijul_channel` now. I don't know which method is preferred so I first committed a set of tests which initializes a Fossil repository+checkout and calls out to the fossil binary, and then a commit which alters the tests to use command mocking. The 9c52cb1c600dda2cd42e094f4577bdfb334b4807-commit can be reverted if calling the binary is preferred.

File: src/test/mod.rs [hunk 1]
```diff
@@ -167,13 +167,22 @@ impl<'a> ModuleRenderer<'a> {
...
 }
 
 pub fn fixture_repo(provider: FixtureProvider) -> io::Result<TempDir> {
     match provider {
+        FixtureProvider::Fossil => {
+            let path = tempfile::tempdir()?;
+            fs::OpenOptions::new()
+                .create(true)
+                .write(true)
+                .open(path.path().join(".fslckout"))?;
```
Comment (davidkna) [inline src/test/mod.rs:183]: To reduce intermittent test failures:
```suggestion
                .open(path.path().join(".fslckout"))?
                .sync_all()?;
```

File: src/modules/fossil_branch.rs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let is_checkout = context
+        .try_begin_scan()?
+        .set_files(&[".fslckout"])
+        .is_match();
+
+    if !is_checkout {
+        return None;
+    }
+
```
Comment (davidkna) [inline src/modules/fossil_branch.rs:19]: Please move this after the `config.disabled` check, because it should be cheaper.

**Review by davidkna — CHANGES_REQUESTED** [review]

File: src/modules/fossil_branch.rs (hunk 2 above)
Comment (VegardSkui) [reply src/modules/fossil_branch.rs]: Changed the order of the checks now. This should probably be changed in the `pijul_channel` module as well to keep consistency between the modules.

**Review by davidkna — APPROVED** [review]
LGTM

**Comment by davidkna** [comment]
Thanks for the contribution @VegardSkui!

**Pull request merged**